### ISSUE TITLE
multiple interpreters to sandbox jar imports?

### ISSUE THREAD

**Issue by nimbusgo** [ibm-et/spark-kernel#131]
I'm looking for a way to emulate a spark jobserver style pattern, where you have a jar, you load/import it and run some main or function from it. Later, this terminates, and you have an updated version of the same jar that you want to run now. Is it possible to do this without restarting the kernel, and hopefully without restarting the spark context?

I was poking through the source, and I noticed that the KernelBootstrap holds the sparkContext and is able to hold multiple interpreters. It appears to only ever hold one interpreter in a Seq, but I'm wondering if this could be adjusted. The interpreter is eventually invoked through an actor, right? Can there be multiple concurrent InterpreterActors, or can you restart an Interpreter?

I'd like to basically be able to have multiple repl sessions that don't interact except for sharing the same sparkContext. It would be nice if one could initiate a new interpreter, import a bunch of dependencies, do some things, and then discard this repl session and start a new one. Is this possible or are there some hurdles with the way spark loads classes that would prevent this?

**Comment by lbustelo** [comment]
Not sure it would be necessary to have separate repls. I think this might work with a combination of kernel.magic.addjar and code to invoke main. When done, run it again. The addjar can be forced to re download the jar if necessary. 

Gino B.

**Comment by nimbusgo** [comment]
This doesn't appear to work without restarting the spark kernel.

See this gist for an example: https://gist.github.com/nimbusgo/f3cd0a050341486364ae

I know that hotswapping jars is asking for trouble and I didn't really expect this to work, that's why I'm looking to work around this issue.

**Comment by rcsenkbeil** [comment]
#### Akka remote actors with individual interpreters ####

Using Akka, one thought would be to run each interpreter in a separate JVM process that communicates back to the main JVM holding the Spark Context. The one gotcha is to ensure that the classes generated by each of the interpreters are stored in the same generated directory that is served by the REPL http server (that's how class files are sent to the cluster).

#### Akka remote actors to purely execute jars ####

Another thought would be to launch the jar itself as a separate JVM process, once again communicating back to the main JVM containing the Spark Context whenever it is accessed. Thereby avoiding the need to reload interpreters.

#### Inline jar loading and executing without polluting interpreter ####

Another option not needing additional JVM processes would be to load the jar (not using addjar) and reflectively executing a method that can receive the Spark Context, passing the kernel's context into it. You can reload classes in the JVM; so, you should be able to reload all of the classes in a jar. You want to avoid going through the Scala interpreter when doing this as it can get pretty hairy.

```scala
/**
 * Evaluates a method from a class within a specified jar, providing it a SparkContext as its parameter.
 *
 * @param path The path to the jar to evaluate
 * @param className The name of the class to instantiate (using empty constructor)
 * @param methodName The name of the method to invoke
 * @param sparkContext The SparkContext instance to pass as the single argument to the method
 *
 * @return The result of executing the method of the class from the jar
 */
def evaluateJar(path: String, className: String, methodName: String, sparkContext: org.apache.spark.SparkContext): Any = {
  val jarClassLoader = new URLClassLoader(Array(new java.io.File(path).toURI.toURL))
  val klass = Class.forName(className, true, jarClassLoader)
  val method = klass.getMethod(methodName, classOf[org.apache.spark.SparkContext])
  val instance = klass.newInstance()
  method.invoke(instance, sparkContext)
}
```

First jar code (copied to `/tmp/tests/testjar.jar`):
<img width="756" alt="screen shot 2015-09-14 at 3 12 59 pm" src="https://cloud.githubusercontent.com/assets/2481802/9860711/5a917956-5af3-11e5-89ac-14e2042f2fbe.png">

Second jar code (copied to `/tmp/tests/testjar2.jar`):
<img width="771" alt="screen shot 2015-09-14 at 3 13 10 pm" src="https://cloud.githubusercontent.com/assets/2481802/9860719/72281994-5af3-11e5-83c6-98e2d6ad30e5.png">

Executing these jars:
<img width="1125" alt="screen shot 2015-09-14 at 3 13 54 pm" src="https://cloud.githubusercontent.com/assets/2481802/9860726/7b2b0268-5af3-11e5-8d00-696afbbe59db.png">

Notice how the interpreter is not polluted with the jar classes as they use a specific classloader that goes out of scope once the code has finished executing.

**Comment by nimbusgo** [comment]
This looks great! I'm looking into eventually doing something interesting with the classloaders, but this will do fine for now. This looks like it will enable a pattern where we can initialize with some jars which aren't changing, then developers can execute short lived jars that might change without polluting the interpreter.

Here's a question, it appears that there aren't problems with passing in the spark context and doing something with it. But what if you're passing in a reference to the kernel and registering msg handlers and such. Do you think there might be any issues with that or should that be fine?

**Comment by rcsenkbeil** [comment]
By _msg handlers_, are you referring to Comm handlers? Or injecting new actors? Or something else? I don't foresee any issues with registering Comm from other jars after passing in the kernel api object. In fact, we've done that in the past via a magic that would take the kernel object and register custom Comm open/msg/close messages.

If you were referring to actors, we don't have any system in place to add your own custom actors, although that doesn't mean that it isn't possible to inject actors into the actor system to serve your own purposes. It might be interesting to open up the possibility of allowing the paths of actors, which you can find [here](https://github.com/ibm-et/spark-kernel/blob/master/protocol/src/main/scala/com/ibm/spark/kernel/protocol/v5/package.scala#L136), to be modified to point to other actors.

**Comment by nimbusgo** [comment]
Yes, comm handlers. Glad to hear you guys have done that before.
